Simple Draw.io Integration

zainkhatri opened the pull request:
# fmdtools DrawIO Integration

## What We Added

### New Files
None - We removed all complex files that @hulsed proposed

### Modified Files
- `fmdtools/define/architecture/base.py` - Added `as_drawio()` method
- `fmdtools/define/architecture/function.py` - Added `as_drawio()` method  
- `fmdtools/analyze/graph/base.py` - Added clean `_draw_drawio()` method
- `fmdtools/analyze/graph/__init__.py` - Removed drawio import

## What It Does
Converts Python models into Draw.io diagrams with:

- Functions and flows
- Clean, simple layout with 500px spacing (as default)
- Uses existing style system (no new dependencies)

## What We DON'T Support (Yet)
- **Actions** - Not actually rendering in current version
- **Components** - Not actually rendering in current version  
- **Environments** - Not actually rendering in current version
- **Architectures** - Not actually rendering in current version

## How to Use
```python
# Generate DrawIO diagram from any model
model.as_drawio('my_diagram.drawio')
```

Example with Tank model:
```python
from examples.tank.tank_model import Tank
Tank().as_drawio('tank_diagram.drawio')
```

Open the .drawio file in Draw.io.

## Tested Examples
- Tank: `Tank().as_drawio('tank.drawio')`
- Rover: `Rover().as_drawio('rover.drawio')`  
- EPS: `EPS().as_drawio('eps.drawio')`
- Pump: `Pump().as_drawio('pump.drawio')`

All models generate spaced out XML output and load perfectly in Draw.io

## What We Removed (Complex Features Removed (for now))
- Advanced layout algorithms (NEATO, spring, hierarchical)
- Performance optimizations and caching
- Complex styling systems
- Niche features (clustering, search, highlighting)

## Result
**Clean, simple implementation that addresses all @hulsed concerns:**
- All methods under 100 lines
-  Uses existing interfaces  
- No duplication
- Minimal doctests
- More maintainable architecture

hulsed commented on fmdtools/analyze/graph/base.py:
```diff
@@ -345,6 +349,18 @@ def draw(self, figsize=(12, 10), title="", fig=False, ax=False, withlegend=True,
         ax : matplotlib axis
             Ax in the figure
         """
+        if format == 'drawio':
+            return self._draw_drawio(saveas=saveas, **kwargs)
+        else:
+            return self._draw_matplotlib(figsize=figsize, title=title, fig=fig, ax=ax, 
+                                       withlegend=withlegend, legend_bbox=legend_bbox, 
+                                       legend_loc=legend_loc, legend_labelspacing=legend_labelspacing,
+                                       legend_borderpad=legend_borderpad, saveas=saveas, **kwargs)
+
+    def _draw_matplotlib(self, figsize=(12, 10), title="", fig=False, ax=False, withlegend=True,
+                         legend_bbox=(1, 0.5), legend_loc="center left", legend_labelspacing=2,
+                         legend_borderpad=1, saveas='', **kwargs):
+        """Draw graph using matplotlib (original draw method)."""
```
Keep Graph.draw() just for working with networkx and put this in draw_drawio.

For some context, we previously had a "render" argument that would switch between drawing with graphviz and drawing with networkx/matplotlib. But, because most of the keyword arguments weren't shared between both renderers (because they have different interfaces) we split them into two so we could document the arguments individually. So instead we now just have a draw_graphviz method (with its own documented arguments/usage) and leave draw() for working with networkx/matplotlib. So to keep it consistent, just have a draw_drawio--no need to provide a top-level interface for this.

hulsed commented on fmdtools/analyze/graph/base.py:
```diff
@@ -370,6 +386,246 @@ def draw(self, figsize=(12, 10), title="", fig=False, ax=False, withlegend=True,
                        loc=legend_loc, add_handles=edge_handles)
         return fig, ax
 
+    def _draw_drawio(self, saveas='', **kwargs):
+        """Generate DrawIO diagram using existing graph structure and positions."""
```
Per above, make this a non-private method and document arguments/usage.

hulsed commented on fmdtools/analyze/graph/base.py:
```diff
@@ -370,6 +386,246 @@ def draw(self, figsize=(12, 10), title="", fig=False, ax=False, withlegend=True,
+            return
+        
+        # Find the range of positions
+        x_coords = [pos[0] for pos in self.pos.values() if hasattr(pos, '__len__') and len(pos) >= 2]
+        y_coords = [pos[1] for pos in self.pos.values() if hasattr(pos, '__len__') and len(pos) >= 2]
```
What is the case where pos doesn't have __len__? Maybe that data structure should just be converted to something usable?

zainkhatri replied:
The unnecessary hasattr(pos, '__len__') and len(pos) >= 2 checks were taken out. The code now accesses pos[0] and pos[1] directly since set_pos() guarantees positions are always (x, y) tuples.

hulsed commented on fmdtools/analyze/graph/base.py:
```diff
@@ -820,7 +1102,7 @@ def data_error(data, average):
         q3.append(np.percentile(current_array, 75))
     lower_error = [x - y for x, y in zip(average, q1)]
-    upper_error = [x - y for x, y in zip(q3, average)]
+    upper_error = [x - y for x, y in zip(average, q3)]
```
The previous was correct here as far as I can tell. If this is confusing, you can re-write both for clarity:

lower_error = [avg - low for avg, low in zip(average, q1)]
upper_error = [hi - avg for avg, hi in zip(average, q1)]

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -30,10 +30,10 @@
 
-The “"Fault Model Design tools - fmdtools version 2"” software is licensed
+The ""Fault Model Design tools - fmdtools version 2"" software is licensed
 under the Apache License, Version 2.0 (the "License"); you may not use this
 file except in compliance with the License. You may obtain a copy of the
-License at http://www.apache.org/licenses/LICENSE-2.0. 
+License at http://www.apache.org/licenses/LICENSE/2.0. 
```
Revert

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -116,6 +116,23 @@ def nx_kwargs(self):
+            if k.startswith('drawio_'):
+                kwargs[k[7:]] = v  # Fixed: 'drawio_' is 7 characters, not 8
+        
+        # Phase 4: Add advanced styling options
+        if hasattr(self, 'drawio_gradient') and self.drawio_gradient:
+            kwargs['gradient'] = True
+        if hasattr(self, 'drawio_shadow') and self.drawio_shadow:
+            kwargs['shadow'] = True
+        if hasattr(self, 'drawio_rounded') and self.drawio_rounded:
+            kwargs['rounded'] = True
```
Isn't this redundant with what you already have above?

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -137,25 +154,56 @@ class EdgeStyle(BaseStyle):
-    gv_arrowhead: ClassVar[str] = 'open'
-    gv_color: ClassVar[str] = 'black'
-    gv_style: ClassVar[str] = 'solid'
+    nx_edge_color: str = 'black'
+    nx_style: str = 'solid'
+    nx_arrows: bool = False
+    nx_arrowstyle: str = '-|>'
+    nx_arrowsize: int = 15
+    gv_arrowhead: str = 'open'
+    gv_arrowtail: str = 'none'
+    gv_dir: str = 'forward'
+    gv_color: str = 'black'
+    gv_style: str = 'solid'
+    drawio_strokewidth: int = 1
+    drawio_strokecolor: str = 'white'  # Changed from 'black' to 'white'
+    drawio_startarrow: str = 'none'
+    drawio_endarrow: str = 'none'
+    drawio_dashed: bool = False
```
If you set these as class variables (using the syntax that was here before), you can set them as fields with defaults in the subclasses (also as was done before).

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -357,30 +417,60 @@ class NodeStyle(BaseStyle):
                      static=dict(nx_node_color='cyan', gv_fillcolor='cyan'),
                      dynamic=dict(nx_edgecolors='teal', gv_color='teal'))
 
+    def __init__(self, *args, styles={}, **kwargs):
+        super().__init__(*args, styles=styles, **kwargs)
+        # Set DrawIO defaults based on node type
+        self._set_drawio_defaults()
+
+    def _set_drawio_defaults(self):
+        """Set DrawIO defaults based on the node type."""
+        # This will be overridden by subclasses if needed
+        pass
+
```
Again this shouldn't be necessary given how the inheritance is set up here

zainkhatri replied:
Yeah, the_set_drawio_defaults methods are gone. Style attributes are now set directly in the __init__ methods.

hulsed commented on fmdtools/define/architecture/action.py:
```diff
@@ -406,6 +406,32 @@ def as_modelgraph(self, gtype=ActionArchitectureGraph, **kwargs):
         """Create and return the corresponding ModelGraph for the Object."""
         return gtype(self, **kwargs)
 
+    def as_drawio(self, saveas='', gtype=ActionArchitectureGraph, **kwargs):
+        """
+        Generate DrawIO diagram from the action architecture.
+        
+        Parameters
+        ----------
+        saveas : str, optional
+            File path to save the DrawIO XML. If empty, returns XML content.
+        gtype : class, optional
+            Graph type to use for visualization. Default is ActionArchitectureGraph.
+        **kwargs : dict
+            Additional arguments passed to the graph creation.
+            
+        Returns
+        -------
+        str
+            DrawIO XML content if saveas is empty, otherwise None.
+            
+        Examples
+        --------
+        >>> xml_content = arch.as_drawio()  # Get XML content
+        >>> arch.as_drawio("action_architecture.drawio")  # Save to file
+        """
+        graph = self.as_modelgraph(gtype=gtype, **kwargs)
+        return graph.draw(saveas=saveas, format='drawio', **kwargs)
+
```
Given that as_modelgraph already changes the graph type for each architecture, this method only needs to be defined for the Architecture class.

hulsed commented on fmdtools/define/architecture/base.py:
```diff
@@ -381,6 +381,32 @@ def as_modelgraph(self, gtype=ArchitectureGraph, **kwargs):
         """Create and return the corresponding ModelGraph for the Object."""
         return gtype(self, **kwargs)
 
+    def as_drawio(self, saveas='', gtype=ArchitectureGraph, **kwargs):
```
you can just have gtype be one of the kwargs since the default will already change in as_modelgraph

zainkhatri replied:
Yeah, the parameter is gone and passes kwargs through to as_modelgraph()

hulsed reviewed (changes requested):
Looks pretty good--see comments for some changes to make

hulsed commented:
@zainkhatri as you fix these make sure to respond to the comments so i know what is addressed/not addressed

zainkhatri commented on fmdtools/analyze/graph/__init__.py:
```diff
@@ -6,10 +6,10 @@
 
-The “"Fault Model Design tools - fmdtools version 2"” software is licensed
+The ""Fault Model Design tools - fmdtools version 2"" software is licensed
 under the Apache License, Version 2.0 (the "License"); you may not use this
 file except in compliance with the License. You may obtain a copy of the
-License at http://www.apache.org/licenses/LICENSE-2.0. 
+License at http://www.apache.org/licenses/LICENSE/2.0. 
```
The format parameter has been removed from Graph.draw(). It now only handles matplotlib rendering and calls _draw_matplotlib() directly

zainkhatri commented:
@hulsed I've responded to your comments and implemented the requested changes. Sorry for the inconvenience with the initial implementation. All fixes are now committed and ready for your review :)

hulsed commented:
@zainkhatri one last thing to fix--preferably, BaseStyle and its subclasses should be dataclasses as they were before. At the very least, if you are changing BaseStyle and its subclasses to be standard object classes instead of a dataclass, you need to make it so that the external behavior when called Graph class is the same such that you can change aspects of the style with set_edge_styles and set_node_styles via edge_style_factory and node_style_factory.

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -326,15 +340,15 @@ def edge_style_factory(style_tag, styles={}, **kwargs):
-    >>> sv = a_s.show_gv(disp=False, saveas=loc+'gv/activation.svg')
+    >>> sv = sv = fs.show_gv(disp=False, saveas=loc+'gv/activation.svg')
 
     >>> c_s = edge_style_factory('containment')
     >>> fig, ax = c_s.show_nx(saveas=loc+'nx/containment.svg')
-    >>> sv = c_s.show_gv(disp=False, saveas=loc+'gv/containment.svg')
+    >>> sv = sv = fs.show_gv(disp=False, saveas=loc+'gv/containment.svg')
 
     >>> cs = edge_style_factory('connection')
     >>> fig, ax = cs.show_nx(saveas=loc+'nx/connection.svg')
-    >>> sv = cs.show_gv(disp=False, saveas=loc+'gv/connection.svg')
+    >>> sv = sv = fs.show_gv(disp=False, saveas=loc+'gv/connection.svg')
```
These tests shouldn't have been modified

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -620,19 +658,19 @@ def node_style_factory(style_tag, styles={}, **kwargs):
     >>> a_s = node_style_factory('architecture')
     >>> fig, ax = a_s.show_nx(saveas=loc+'nx/architecture.svg')
-    >>> sv = a_s.show_gv(disp=False, saveas=loc+'gv/architecture.svg')
+    >>> sv = sv = fs.show_gv(disp=False, saveas=loc+'gv/action.svg')
```
These tests shouldn't have been modified

hulsed commented on fmdtools/analyze/graph/style.py:
```diff
@@ -528,64 +548,82 @@ def __init__(self, *args, **kwargs):
-    gv_style: str = 'filled'
-    gv_shape: str = 'pentagon'
-    gv_penwidth: str = '0'
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Override specific values in __init__ to avoid field conflicts
+        self.nx_node_shape = 'p'
+        self.nx_linewidths = 0
+        self.gv_style = 'filled'
+        self.gv_shape = 'pentagon'
+        self.gv_penwidth = '0'
```
This still has the problem that is that arguments passed by the user are going to be overwritten by these defaults. The defaults should be overwritten by user arguments instead

hulsed reviewed (commented):
See comments. There are still problems with the Style classes in that the defaults are going to overwrite user keyword arguments

zainkhatri commented on fmdtools/analyze/graph/style.py:
```diff
@@ -176,33 +167,26 @@ class EdgeStyle(BaseStyle):
+        self.drawio_endarrow = 'none'
+        self.drawio_dashed = False
+        self.modifiers = dict(degraded=dict(nx_edge_color='orange', gv_color='orange'),
+                             active=dict(nx_edge_color='green', gv_color='green'))
```
I moved the defaults back into dataclass fields and updated the methods to use asdict(self), so user kwargs now override defaults and modifiers apply last.

zainkhatri commented:
@hulsed I converted the style classes back into dataclasses and adjusted how kwargs are handled so that user supplied arguments take precedence over defaults. I added a dedicated drawio_kwargs() method to better support DrawIO styling, and fixed small issues like suppressing arrow kwargs when nx_arrows=False.

hulsed reviewed (approved):
Looks good
